tidy3d/components/parameter_perturbation.py

```diff
@@ -0,0 +1,722 @@
+        """Sample perturbation.
+
+        Parameters
+        ----------
+        temperature : Union[ArrayLike[float], SpatialDataArray]
+            Temperature sample point(s).
+
+        Returns
+        -------
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]
+            Sampled perturbation value(s).
+        """
+
+        return self.coeff * (temperature - self.temperature_ref)
+
+    @cached_property
+    def is_complex(self) -> bool:
+        """Whether perturbation is complex valued."""
+        return np.iscomplex(self.coeff)
+
+
+class CustomHeatPerturbation(HeatPerturbation):
+    """Custom heat perturbation."""
+
+    perturbation_values: HeatDataArray = pd.Field(
+        ...,
+        title="Perturbation Values",
+        description="Sampled perturbation values.",
+    )
+
+    temperature_range: Tuple[pd.NonNegativeFloat, pd.NonNegativeFloat] = pd.Field(
+        None,
+        title="Temperature range",
+        description="Temparature range in which perturbation model is valid. For "
+            ":class:`.CustomHeatPerturbation` this field is computed automatically based on "
+            "provided ``perturbation_values``",
+        units=KELVIN,
+    )
+
+    @cached_property
+    def perturbation_range(self) -> Union[Tuple[float, float], Tuple[Complex, Complex]]:
+        """Perturbation range."""
+        return np.min(self.perturbation_values), np.max(self.perturbation_values)
+
+    @pd.root_validator(skip_on_failure=True)
+    def compute_temperature_range(cls, values):
+        """Compute and set temperature range based on provided ``perturbation_values``."""
+        if values["temperature_range"] is not None:
+            log.warning(
+                "Temperature range for 'CustomHeatPerturbation' is calculated automatically "
+                "based on provided 'perturbation_values'. Provided 'temperature_range' will be overwritten."
+            )
+
+        perturbation_values = values["perturbation_values"]
+
+        temperature_range = (
+            float(np.min(perturbation_values.coords["T"])),
+            float(np.max(perturbation_values.coords["T"])),
+        )
+
+        values.update({"temperature_range": temperature_range})
+
+        return values
+
+    @ensure_temp_in_range
+    def sample(
+        self, temperature: Union[ArrayLike[float], SpatialDataArray]
+    ) -> Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]:
+        """Sample perturbation.
+
+        Parameters
+        ----------
+        temperature : Union[ArrayLike[float], SpatialDataArray]
+            Temperature sample point(s).
+
+        Returns
+        -------
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]
+            Sampled perturbation value(s).
+        """
+
+        t_range = self.temperature_range
+        temperature_clip = np.clip(temperature, t_range[0], t_range[1])
+        return self.perturbation_values.interp(T=temperature_clip).drop_vars("T")
```

Comment:
shall we have "linear" and "nearest" option here?

Reply:
yeah, added for completeness